icelake/src/io/appender/base_file_writer.rs

```diff
@@ -0,0 +1,377 @@
+//! A module provide `RollingWriter`.
+use arrow_array::RecordBatch;
+use arrow_cast::cast;
+use async_trait::async_trait;
+use std::sync::Arc;
+
+use crate::config::RollingWriterConfig;
+use crate::io::location_generator::FileLocationGenerator;
+use crate::io::{
+    FileWriteResult, FileWriter, FileWriterBuilder, IcebergWriteResult, IcebergWriter,
+    IcebergWriterBuilder, SingletonWriter,
+};
+
+use crate::{Error, Result};
+use arrow_schema::{DataType, SchemaRef};
+
+#[derive(Clone)]
+pub struct BaseFileWriterBuilder<B: FileWriterBuilder> {
+    location_generator: Arc<FileLocationGenerator>,
+    rolling_config: Option<RollingWriterConfig>,
+    inner: B,
+}
+
+impl<B: FileWriterBuilder> BaseFileWriterBuilder<B> {
+    pub fn new(
+        location_generator: Arc<FileLocationGenerator>,
+        rolling_config: Option<RollingWriterConfig>,
+        inner: B,
+    ) -> Self {
+        Self {
+            location_generator,
+            rolling_config,
+            inner,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: FileWriterBuilder> IcebergWriterBuilder for BaseFileWriterBuilder<B>
```

Comment:
Why this is a `IcebergWriter`? I think rolling writer should belong to file writer layer?

Reply:
For now, our arch look like this, and that's why BaseFileWriter is a IcebergWriter rather than FileWriter. BaseFileWriter like the writer bridge IcebergWriter and FileWriter.
![image](https://github.com/icelake-io/icelake/assets/43447882/0d0037ea-676d-47a5-876c-f6347c6db1f4)

I try to convert it to FileWriter but find some problem:
We can see that FileWriter doesn't have flush, it only has close because it don't need to be a Box object so close interface is simpler.
```
trait FileWriter {
  fn close(self)
}

trait IcebergWriter {
  fn flush(&mut self)
}
```
This means that the IcebergWriter use FileWriter should implement the flush using close, e.g. in BaseFileWriter, we close and then reopen a new writer.
```
    async fn flush(&mut self) -> Result<Self::R> {
        self.close_current_writer().await?;
        self.open_new_writer().await?;
        Ok(self.result.flush())
    }
```
So, if we let BaseFileWriter be the file writer, this job needs to be implemented in DataFileWriter, PositionDeleteWriter, EqualityDeletetWriter 3 times. Making the BaseFileWriter be IcebergWriter can let it decouple the FileWriter and IcebergWriter, if the interface of FileWriter change, what we need to change is only the BaseFileWriter.

Reply:
Good abstraction leads to clean reuse of different components, but purely saving code should not be the goal of abstraction.

Reply:
Got it. Thanks! Let's move!